src/gui/Layout.h

```diff
@@ -107,45 +107,40 @@ class Layout
 
 private:
 	void checkScroll(GtkAdjustment* adjustment, double& lastScroll);
+
 	void setLayoutSize(int width, int height);
 
 private:
 	XOJ_TYPE_ATTRIB;
 
-	XournalView* view = NULL;	
-
 	LayoutMapper mapper;
 
-	ScrollHandling* scrollHandling = NULL;
+	XournalView* view = nullptr;
+	ScrollHandling* scrollHandling = nullptr;
+
+	std::vector<int> widthCols;
+	std::vector<int> heightRows;
 
 	double lastScrollHorizontal = -1;
 	double lastScrollVertical = -1;
 
 	/**
-	 * The last width of the widget
+	 * The last width and height of the widget
 	 */
 	int lastWidgetWidth = 0;
+	int lastWidgetHeight = 0;
 
 	/**
 	 * The width and height of all our pages
 	 */
-	int layoutWidth = 0;
-	int layoutHeight = 0;
-	
-	
-	/**
-	 *The following are useful for locating page at a pixel location
-	 */
-	
-	int rows;
-	int columns;
-	
-	std::vector<int> sizeCol;
-	std::vector<int> sizeRow;
-	
+	int minWidth = 0;
+	int minHeight = 0;
+
 	/**
 	 * cache the last GetViewAt() row and column.
 	 */
 	int lastGetViewAtRow = 0;
 	int lastGetViewAtCol = 0;
+
+	bool valid = false;
```

Comment:
Could you add a comment what this does?

Reply:
Done, but i would rather remove it completely. (Part of a big refactoring / completely unrelated to zoom). This is just a workaround.

Reply:
If you can remove it without influencing the behaviour, just go for it. It is the main goal to decrease complexity of the code after all.

Reply:
Unfortunately i would have to rewrite Layout::getViewAt and Layout::updateVisibility, that was a too big change for me, i also have other things to do in my private time. :D. But I added some todos, to indicate, that it could/should be refactored away.

Reply:
Ahh if it is marked that is already good. It can still be done in a separate PR. If you might have the time later on you can still try your hand. The work you provide is very good so far.